### PR TITLE
Feature: possibility to use placeholder icon in legend for raster layer. funded by [Canton of Glarus] [https://www.gl.ch/verwaltung/bau-und-umwelt/hochbau/raumentwicklung-und-geoinformation/geoportal-kanton-glarus.html/808]

### PR DESCRIPTION
**Pull request by mhugent**
We use some paletted raster layers as background maps with QGIS Server. Currently, the raster legend shows all the palette entries with color and index number. While this might be fine in certain use-cases, the legend with 256 entries for a color background map is too much in our case (and the index number does not bring useful information for the user). Therefore this PR adds the feature to select a legend placeholder image for a raster or vector layer.

legend_placeholder_image

### CODE REVIEW

**Comment by gioman** [comment]
> Or maybe even to let the user select any icon to show for the layer?

@mhugent this would be highly appreciated: both Geoserver and Mapserver allow to use a custom image for a WMS layer legend.

**Comment by DelazJ** [comment]
@mhugent for the documentation purpose, can you add a screenshot with the feature in place? Thanks.

File: src/core/raster/qgsrasterlayer.h [hunk 1]
```diff
@@ -467,6 +467,15 @@ class CORE_EXPORT QgsRasterLayer : public QgsMapLayer
 
     QgsMapLayerTemporalProperties *temporalProperties() override;
 
+    //! True if a placeholder legend is shown in the legend rather than all the color entries
```
Comment (nyalldawson) [inline src/core/raster/qgsrasterlayer.h:470]: Don't forget \since annotation

**Comment by nyalldawson** [comment]
> Or maybe even to let the user select any icon to show for the layer?

This sounds ideal to me!

File: src/core/qgsmaplayer.h [hunk 2]
```diff
@@ -1264,6 +1264,19 @@ class CORE_EXPORT QgsMapLayer : public QObject
      */
     virtual QgsMapLayerElevationProperties *elevationProperties() { return nullptr; }
 
+    /**
+     * Returns path to the placeholder image or an empty string if a generated legend is shown
+     * \return placholder image path
```
Comment (nyalldawson) [inline src/core/qgsmaplayer.h:1269]: Also needs since

File: src/core/qgsmaplayerlegend.cpp [hunk 3]
```diff
@@ -359,6 +360,19 @@ QList<QgsLayerTreeModelLegendNode *> QgsDefaultVectorLayerLegend::createLayerTre
 {
   QList<QgsLayerTreeModelLegendNode *> nodes;
 
+  if ( mLayer )
+  {
+    QString placeholderImage = mLayer->legendPlaceholderImage();
+    if ( !placeholderImage.isEmpty() )
+    {
+      QgsImageCache ic;
```
Comment (nyalldawson) [inline src/core/qgsmaplayerlegend.cpp:368]: This should use the image cache instance attached to QgsApplication

File: src/core/qgsmaplayerlegend.cpp [hunk 4]
```diff
@@ -502,7 +516,15 @@ QList<QgsLayerTreeModelLegendNode *> QgsDefaultRasterLayerLegend::createLayerTre
     nodes << new QgsWmsLegendNode( nodeLayer );
   }
 
-  if ( mLayer->renderer() )
+  QString placeholderImage = mLayer->legendPlaceholderImage();
+  if ( !placeholderImage.isEmpty() )
+  {
+    QgsImageCache ic;
```
Comment (nyalldawson) [inline src/core/qgsmaplayerlegend.cpp:522]: Same here

**Comment by nyalldawson** [comment]
Looking good! Just needs some tests now...

File: src/core/qgsmaplayerlegend.cpp (hunk 3 above)
Comment (mhugent) [reply src/core/qgsmaplayerlegend.cpp]: Ah cool, didn't know there is an image cache instance in QgsApplication

**Comment by mhugent** [comment]
@DelazJ : here a screenshot:

legend_placeholder_image

**Comment by mhugent** [comment]
Unit test is going to come...

**Comment by DelazJ** [comment]
> @DelazJ : here a screenshot:
Thanks @mhugent. Can you (or someone that has the rights) please add it to the PR first message so that it is grabbed by the bots?

**Review by nyalldawson — APPROVED** [review]
LGTM! Merge away when the tests are ready!

**Pull request merged**